Fix the test failure on volume snapshot

humblec opened the pull request:
Since we have upgraded the snapshot controller version to v6, the
snapshot tests looks to be failing in the testgrid. It has been mainly
because the latest version of snapshot controller stopped serving
v1beta1 APIs. The sidecar image versions in the tests also has to be
updated to make sure these are compatible.

Additional Ref#
https://github.com/kubernetes-csi/external-snapshotter/issues/736
https://testgrid.k8s.io/sig-storage-kubernetes#volume-snapshot

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

/kind cleanup

```release-note
NONE
```

humblec commented:
/sig storage

humblec commented:
Cc @mattcary @xing-yang

Abirdcfly commented:
Hi @humblec will this change fix https://github.com/kubernetes/kubernetes/issues/110658 by the way?

humblec commented:
> Hi @humblec will this change fix #110658 by the way?

@Abirdcfly at glance on the build logs, it looks like this can fix above one.

mattcary commented:
Looks like the same failures? eg for the pd csi should provision with snapshot data source, the driver & sidecars successfully mark the snapshot completed at 9:26, but it's still not being seen as ready by the time it times out at 9:31.

Unfortunately it doesn't look like the snapshot controller logs get in the artifacts (?) so I'm not sure what's going on from its end.

xing-yang commented:
We missed a RBAC:

```
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/110826/pull-kubernetes-e2e-gce-storage-snapshot/1541827767174172672/build-log.txt


I0628 17:40:32.728] Jun 28 17:40:32.702: INFO: At 2022-06-28 17:30:26 +0000 UTC - event for snapshot-j2jrs: {snapshot-controller } SnapshotFinalizerError: Failed to check and update snapshot: snapshot controller failed to update snapshottable-stress-6439/snapshot-j2jrs on API server: volumesnapshots.snapshot.storage.k8s.io "snapshot-j2jrs" is forbidden: User "system:serviceaccount:kube-system:volume-snapshot-controller" cannot patch resource "volumesnapshots" in API group "snapshot.storage.k8s.io" in the namespace "snapshottable-stress-6439"
```

xing-yang commented:
A few "patch" rules are missing:

https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml

https://github.com/kubernetes/kubernetes/blob/v1.25.0-alpha.2/cluster/addons/volumesnapshots/volume-snapshot-controller/rbac-volume-snapshot-controller.yaml

drgnchan commented on cluster/addons/volumesnapshots/volume-snapshot-controller/rbac-volume-snapshot-controller.yaml:
```diff
@@ -35,13 +35,16 @@ rules:
   - apiGroups: ["snapshot.storage.k8s.io"]
     resources: ["volumesnapshotcontents"]
-    verbs: ["create", "get", "list", "watch", "update", "delete"]
+    verbs: ["create", "get", "list", "watch", "update", "delete" "patch"]
```
is there lacking `,` after `"delete"`

humblec replied:
it was an intermediate push.. thanks

humblec commented:
/test pull-kubernetes-unit

humblec commented:
The finalizerRBAC error does not pop anymore , however the latest errors seen  in the test run 
```
 I0629 05:26:36.871] Jun 29 05:26:36.869: INFO: received snapshotStatus map[boundVolumeSnapshotContentName:snapcontent-0a1e6840-eee5-4983-b0fd-340776ce22f8 creationTime:2022-06-29T05:21:30Z error:map[message:Failed to check and update snapshot content: failed to take snapshot of the volume projects/k8s-gce-gci-1-6-mstr-ctl-skew/zones/us-west1-b/disks/pvc-cd97bddd-38dc-4329-a6cb-b0ee8c23c96d: "rpc error: code = Aborted desc = An operation with the given Volume ID projects/k8s-gce-gci-1-6-mstr-ctl-skew/zones/us-west1-b/disks/pvc-cd97bddd-38dc-4329-a6cb-b0ee8c23c96d already exists" time:2022-06-29T05:21:36Z] readyToUse:false restoreSize:5Gi] 
....

I0629 05:27:55.258] Jun 29 05:27:55.217: INFO: At 2022-06-29 05:20:59 +0000 UTC - event for pod-44933fe1-9610-4f65-a812-0985411373cc: {kubelet e2e-45301dc696-30ee2-minion-group-5gsc} FailedMount: MountVolume.SetUp failed for volume "kube-api-access-7c82z" : failed to sync configmap cache: timed out waiting for the condition `
....

 I0629 05:27:55.337] Jun 29 05:27:55.242: INFO: At 2022-06-29 05:27:07 +0000 UTC - event for snapshot-k97zx: {snapshot-controller } SnapshotContentObjectDeleteError: Failed to delete snapshot content API object
I0629 05:27:55.337] Jun 29 05:27:55.248: INFO: At 2022-06-29 05:27:07 +0000 UTC - event for snapshot-skxtj: {snapshot-controller } SnapshotContentObjectDeleteError: Failed to delete snapshot content API object
I0629 05:27:55.337] Jun 29 05:27:55.248: INFO: At 2022-06-29 05:27:08 +0000 UTC - event for snapshot-zm47v: {snapshot-controller } SnapshotContentObjectDeleteError: Failed to delete snapshot content API object 
...
 
```

@mattcary I am wondering this has something to do with the provider side ( gcp drivers) snapshot stress test .

humblec committed:
Fix the test failure on volume snapshot

Since we have upgraded to snapshot controller version to v6, the
snapshot tests looks to be failing in the testgrid. It has been mainly
because the latest version of snapshot controller stopped serving
v1beta1 APIs. The sidecar image versions in the tests also has to be
updated to make sure these are compatible.

This commit add missing RBAC rules for the controller as per the
latest version.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

xing-yang commented:
> I0629 09:52:51.527] Jun 29 09:52:51.418: INFO: At 2022-06-29 09:51:30 +0000 UTC - event for snapshot-s4l8z: {snapshot-controller } SnapshotFinalizerError: Failed to check and update snapshot: snapshot controller failed to update snapshottable-stress-4983/snapshot-s4l8z on API server: VolumeSnapshot.snapshot.storage.k8s.io "snapshot-s4l8z" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"snapshot.storage.kubernetes.io/volumesnapshot-bound-protection"} 
.....

It seems that a VolumeSnapshot is being deleted (got a deletion timestamp) before it is bound. So it can never be bound. I think the test needs to wait for the VolumeSnapshot to be bound before trying to delete it.

xing-yang commented:
We should probably wait for the VolumeSnapshot to be ready before returning in this function:
https://github.com/kubernetes/kubernetes/blob/v1.25.0-alpha.2/test/e2e/storage/framework/snapshot_resource.go#L77

Otherwise, we need to treat the case where a VolumeSnapshot is being deleted before it is bound as a success case.

xing-yang commented:
It's a race condition.  It could be that the time it takes to finish a snapshot binding is different now.  Take a look of changes in v5 as well.  The new rbac rules were added in v5 when we replaced "update" with "patch" in many places.

humblec commented:
> It's a race condition. It could be that the time it takes to finish a snapshot binding is different now. Take a look of changes in v5 as well. The new rbac rules were added in v5 when we replaced "update" with "patch" in many places.

Yeah, both of the above possibilities are on the same line of my thought. It looks like the snapshot creation is taking more time in the provider for some reason. Also , yes, the update/patch call changes could be triggering it too.

msau42 commented:
@humblec the failing test is for mock driver, not cloud specific

xing-yang commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

xing-yang commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml:
```diff
@@ -39,7 +39,7 @@ spec:
               mountPath: /csi
         - name: csi-provisioner
-          image: registry.k8s.io/sig-storage/csi-provisioner:v3.1.0
+          image: registry.k8s.io/sig-storage/csi-provisioner:v3.2.0
```
The latest is v3.2.1.

humblec commented:
> @humblec the failing test is for mock driver, not cloud specific

@msau42 ah, I was looking at wrong url ( from previous runs ) of logs :(, so the comment. Thanks

humblec commented:
/test pull-kubernetes-e2e-kind-ipv6

leiyiz commented on test/e2e/testing-manifests/storage-csi/mock/csi-mock-driver.yaml:
```diff
@@ -15,7 +15,7 @@ spec:
       containers:
         - name: csi-provisioner
-          image: registry.k8s.io/sig-storage/csi-provisioner:v3.0.0
+          image: registry.k8s.io/sig-storage/csi-provisioner:v3.2.0
```
3.2.1

leiyiz commented:
/retest

humblec commented:
/test pull-kubernetes-unit

msau42 commented:
Are you able to bring up a local environment on a kind cluster?

humblec commented:
Reference# https://github.com/kubernetes/kubernetes/issues/111086#issuecomment-1192166936

Abirdcfly commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

xing-yang commented:
/triage accepted

humblec commented:
Struggling to bring up the kind cluster in my setup, so the delay:

The kind node test image has been built , however while bringing up the cluster locally , the kubelet is failing :

````
I0728 14:31:44.693376     151 manifests.go:154] [control-plane] wrote static Pod manifest for component "kube-controller-manager" to "/etc/kubernetes/manifests/kube-controller-manager.yaml"
I0728 14:31:44.693408     151 manifests.go:99] [control-plane] getting StaticPodSpecs
[control-plane] Creating static Pod manifest for "kube-scheduler"
I0728 14:31:44.693789     151 manifests.go:125] [control-plane] adding volume "kubeconfig" for component "kube-scheduler"
I0728 14:31:44.695556     151 manifests.go:154] [control-plane] wrote static Pod manifest for component "kube-scheduler" to "/etc/kubernetes/manifests/kube-scheduler.yaml"
[etcd] Creating static Pod manifest for local etcd in "/etc/kubernetes/manifests"
I0728 14:31:44.700636     151 local.go:65] [etcd] wrote Static Pod manifest for a local etcd member to "/etc/kubernetes/manifests/etcd.yaml"
I0728 14:31:44.700699     151 waitcontrolplane.go:83] [wait-control-plane] Waiting for the API server to be healthy
I0728 14:31:44.702374     151 loader.go:374] Config loaded from file:  /etc/kubernetes/admin.conf
[wait-control-plane] Waiting for the kubelet to boot up the control plane as static Pods from directory "/etc/kubernetes/manifests". This can take up to 4m0s
...
I0728 14:33:39.206661     151 round_trippers.go:553] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
I0728 14:33:39.706751     151 round_trippers.go:553] GET https://kind-control-plane:6443/healthz?timeout=10s  in 0 milliseconds
[kubelet-check] It seems like the kubelet isn't running or healthy.
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:10248/healthz": dial tcp [::1]:10248: connect: connection refused.

````

not sure what exactly causing the failure ( tried to reduce the cluster node...etc without luck)  .. may be resource issue on local node/setup. still debugging.

dgrisonnet commented:
/remove-sig instrumentation

neolit123 commented:
/remove-area kubeadm
/remove-sig cluster-lifecycle

Abirdcfly commented:
/test pull-kubernetes-e2e-gce-storage-snapshot

xing-yang commented:
Have you rebased?  Can you also update snapshot-controller image to 6.0.1?

humblec commented:
> Have you rebased? Can you also update snapshot-controller image to 6.0.1?

rebased and updated. will check this run.

humblec commented:
/test pull-kubernetes-e2e-kind-ipv6

humblec commented:
/test pull-kubernetes-e2e-gce-100-performance

humblec commented:
/test pull-kubernetes-e2e-kind-ipv6

xing-yang commented:
Very strange.  The test has passed several times on my PR.  Can you squash your commits?

humblec commented:
> Very strange. The test has passed several times on my PR. Can you squash your commits?

Indeed and squashed the commits too..will check how it goes.

humblec commented:
In last commit or in last refresh of this PR I just removed/nullified the snapshotter update on https://github.com/kubernetes/kubernetes/pull/110826/files#diff-1e966d5d795fba59a8ceea256e18d4390e98c4b6fc5adeb135397ab391598174L24 from `5.0.1 to 6.0.1` and all test pass. @xing-yang @msau42 @mattcary

msau42 commented:
@humblec @xing-yang can you clarify what the problem was? Is there some version compatibility between csi-provisioner, csi-snapshotter and snapshot-controller that we need to document?

humblec commented:
> @humblec @xing-yang can you clarify what the problem was? Is there some version incompatibility between csi-provisioner, csi-snapshotter and snapshot-controller that we need to document?

@msau42  While I was updating csi-provisioner , csi-snapshotter and snapshot-controller in all of the test/CI yamls to latest versions `3.2.1` (prov) and `6.0.1` ( snapshot*) `pull-kubernetes-e2e-gce-storage-snapshot` test is failing. 

However If I revert the version to `5.0.1` ( keep this intact) for snapshot controller in  test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml  (https://github.com/kubernetes/kubernetes/pull/110826/files#diff-1e966d5d795fba59a8ceea256e18d4390e98c4b6fc5adeb135397ab391598174R17-L24) and keep all the other yaml updates in place ( as in the current state of this PR) , even this test pass. so, it looks like some thing specific to gce-pd test causing the failure.

humblec commented:
> @humblec @xing-yang can you clarify what the problem was? Is there some version incompatibility between csi-provisioner, csi-snapshotter and snapshot-controller that we need to document?

@msau42  While I was updating csi-provisioner , csi-snapshotter and snapshot-controller in all of the test/CI yamls to latest versions `3.2.1` (prov) and `6.0.1` ( snapshot*) `pull-kubernetes-e2e-gce-storage-snapshot` test is failing. 

However If I revert the version to `5.0.1` ( keep this intact) for snapshot controller in  test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml  (https://github.com/kubernetes/kubernetes/pull/110826/files#diff-1e966d5d795fba59a8ceea256e18d4390e98c4b6fc5adeb135397ab391598174R17-L24) and keep all the other yaml updates in place ( as in the current state of this PR) , even this test pass. so, it looks like some thing specific to gce-pd test causing the failure.

humblec commented:
@xing-yang I think we can take this PR as in this form too. Only *snapshotter* image update is left in the gce ci pod yaml , rest are in place. It may not fix the flakiness in full extent, but  good to have this in too!

msau42 commented:
Ok, can you open a separate bug to track updating pd csi-snapshotter to v6? 

/lgtm
/approve

humblec commented:
> Ok, can you open a separate bug to track updating pd csi-snapshotter to v6?
> 
> /lgtm /approve

Sure , done. https://github.com/kubernetes/kubernetes/issues/112694
